zerver/lib/export.py

```diff
@@ -0,0 +1,653 @@
+    floatify_datetime_fields(response, 'zerver_useractivityinterval', 'end')
+
+    stream_query = Stream.objects.filter(realm=realm)
+    if not include_invite_only:
+        stream_query = stream_query.filter(invite_only=False)
+    response['zerver_stream'] = [model_to_dict(x, exclude=["email_token"]) for x in stream_query]
+    floatify_datetime_fields(response, 'zerver_stream', 'date_created')
+    stream_ids = set(x["id"] for x in response['zerver_stream'])
+    stream_recipients = [model_to_dict(x)
+                         for x in Recipient.objects.filter(type=2, type_id__in=stream_ids)]
+
+    stream_recipient_ids = set(x["id"] for x in stream_recipients)
+
+    stream_subscription_query = Subscription.objects.filter(user_profile__in=user_profile_ids,
+                                                            recipient_id__in=stream_recipient_ids)
+    stream_subscription_dicts = [model_to_dict(x) for x in stream_subscription_query]
+
+    if include_private:
+        # First we get all huddles involving someone in the realm.
+        realm_huddle_subs = Subscription.objects.select_related("recipient").filter(recipient__type=3,
+                                                                                    user_profile__in=user_profile_ids)
+        realm_huddle_recipient_ids = set(sub.recipient_id for sub in realm_huddle_subs)
+
+        # Mark all Huddles whose recipient ID contains a cross-realm user.
+        unsafe_huddle_recipient_ids = set()
+        for sub in Subscription.objects.select_related().filter(recipient__in=realm_huddle_recipient_ids):
+            if sub.user_profile.realm != realm:
+                # In almost every case the other realm will be zulip.com
+                unsafe_huddle_recipient_ids.add(sub.recipient_id)
+
+        # Now filter down to just those huddles that are entirely within the realm.
+        huddle_subs = [sub for sub in realm_huddle_subs if sub.recipient_id not in unsafe_huddle_recipient_ids]
+        huddle_recipient_ids = set(sub.recipient_id for sub in huddle_subs)
+        huddle_ids = set(sub.recipient.type_id for sub in huddle_subs)
+
+        huddle_subscription_dicts = [model_to_dict(x) for x in huddle_subs]
+        huddle_recipients = [model_to_dict(r) for r in Recipient.objects.filter(id__in=huddle_recipient_ids)]
+        response['zerver_huddle'] = [model_to_dict(h) for h in Huddle.objects.filter(id__in=huddle_ids)]
+    else:
+        huddle_recipients = []
+        huddle_subscription_dicts = []
+
+    response["zerver_recipient"] = user_recipients + stream_recipients + huddle_recipients
+    response["zerver_subscription"] = user_subscription_dicts + stream_subscription_dicts + huddle_subscription_dicts
+
+    response["zerver_attachment"] = [model_to_dict(x) for x in
+                                     Attachment.objects.filter(realm=realm)]
+    floatify_datetime_fields(response, 'zerver_attachment', 'create_time')
```

Comment:
This might be overkill, but we could have a method called `fix_fields(response, table)` that consolidates all the fixups into one method that uses some kind of lookup table.  Then somebody can scan the data structure all in one pass to see which fields need time fixups, etc.